File: go/1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

// sum2 expects the slice to be sorted.
// This way, we can find the sum in O(n) with the
// "two-finger" method.
func sum2(arr []int, sum int) (int, int, bool) {
	left, right := 0, len(arr)-1
	for left < right {
		if arr[left]+arr[right] == sum {
			return arr[left], arr[right], true
		}
		if arr[left]+arr[right] < sum {
			left++
		} else {
			right--
		}
	}
	return 0, 0, false
}

// sum3 expects the slice to be sorted.
// We can find the sum of 3 variables in O(n^2).
func sum3(arr []int, sum int) (int, int, int, bool) {
	for _, thd := range arr {
		if fst, snd, ok := sum2(arr, sum-thd); ok {
			return fst, snd, thd, true
		}
	}
	return 0, 0, 0, false
}

func day1() {
	// load the inputs
	file, err := os.Open("./../inputs/1.txt")
	if err != nil {
		fmt.Printf("couldn't open file: %s\n", err.Error())
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)

	expenses := []int{}
	for scanner.Scan() {
		i, err := strconv.Atoi(scanner.Text())
		if err != nil {
			fmt.Printf("couldn't parse input: %s\n", err.Error())
		}
		expenses = append(expenses, i)
	}

	sort.Ints(expenses) // this is O(n*log(n))
	if fst, snd, ok := sum2(expenses, 2020); ok {
		fmt.Printf("%d * %d = %d\n", fst, snd, fst*snd)
	} else {
		fmt.Println("there are no 2 variables that add to 2020")
	}

	if fst, snd, thd, ok := sum3(expenses, 2020); ok {
		fmt.Printf("%d * %d *%d = %d\n", fst, snd, thd, fst*snd*thd)
	} else {
		fmt.Println("there are no 3 variables that add to 2020")
	}
}

File: cpp/1.cpp
#include <string>
#include <algorithm>
#include <fstream>
#include <iostream>
#include <optional>
#include <tuple>
#include <vector>

std::optional<std::tuple<int, int>> sum_two( const std::vector<int>& ints, int sum ) {
    for ( int front = 0, back = ints.size()-1; front < back; ) {
        if ( ints[front] + ints[back] == sum ) return std::tuple<int,int>{ ints[front], ints[back] };
        if ( ints[front] + ints[back] < sum ) front++;
        else back--;
    }
    return std::nullopt;
}

std::optional<std::tuple<int,int,int>> sum_three( const std::vector<int>& ints, int sum ) {
    for ( auto item : ints ) {
        auto result = sum_two( ints, sum - item );
        if ( result.has_value() ) {
            auto [ fst, snd ] = result.value();
            return std::tuple<int, int, int>{ fst, snd, item };
        }
    }
    return std::nullopt;
}

int main() {
    std::fstream file;
    file.open( "./../inputs/1.txt", std::ios::in );

    std::vector<int> ints;
    for ( int item = 0; file >> item; ) ints.emplace_back( item );
    std::sort( ints.begin(), ints.end() );

    auto result_sum_two = sum_two( ints, 2020 );
    if ( result_sum_two.has_value() ) {
        auto [ fst, snd ] = result_sum_two.value();
        std::cout << "sum_two: " << std::to_string(fst * snd) << std::endl;
    } else {
        std::cout << "did not find sum of two for 2020" << std::endl;
    }

    auto result_sum_three = sum_three( ints, 2020 );
    if ( result_sum_three.has_value() ) {
        auto [ fst, snd, thd ] = result_sum_three.value();
        std::cout << "sum_three: " << std::to_string(fst * snd * thd) << std::endl;
    } else {
        std::cout << "did not find sum of three for 2020" << std::endl;
    }
}